### PR TITLE
New scheduler with min-heap, no rAF

### PR DESCRIPTION
**Pull request by 132yse**
This is an important pr, which implements the new scheduler with a new way.

#### Use the min heap sorts priority queue

Give up the linked list and use the min heap. Now it conforms to make the algorithm as possible as standard and easy to understand.

#### Use Message Channel, no requestAnimationFrame.

No need dynamic frame rate adjustment, now requestAnimationFrame is only used for starting a work loop.

### CODE REVIEW

File: src/min-heap.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+function peek (heap) {
+  var first = heap[0]
+  return first || null
+}
+
+const channel = new MessageChannel
+const port = channel.port2
+channel.port1.onmessage = performWork
```
[the remark this answers is not in the document]
Comment (132yse) [reply src/min-heap.js]: It's not equivalent. Message channel happens after browser redpaintinging. Promise happens before it.

[the remark this answers is not in the document]
Comment (mindplay-dk) [reply src/min-heap.js]: Okay. Hmm. But there must be another more appropriate API for this?

Surely this isn't the intended use of an API designed for messaging between windows/frames?

There must be another type of callback that runs at the appropriate time?

[the remark this answers is not in the document]
Comment (132yse) [reply src/min-heap.js]: It seems that there is no API other than setTimeout.
But unfortunately, I found that I couldn't run the use case well with Message Channel.

**Comment by 132yse** [comment]
Unfortunately, I have given the wrong implementation.
The use case is correct. I need to take more time.
So sad……

**Comment by mindplay-dk** [comment]
Yeah, this stuff is *hard*. 😳

Didact successfully did it with `requestIdleCallback` - [see here](https://github.com/pomber/didact/blob/44093cb83f0e20929b4bec0d04baff6bdf3dc155/src/reconciler.js#L30-L56).

I don't know how to compare that with message channel - I can't find any description of when/how message channels get dispatched in the event-loop...

**Comment by 132yse** [comment]
In fact, I also don't know the details. I can only guess and try. I really can't find any documents.

At that time, I guessed that Message Channel was executed after repainting, and the repainting was a high priority task, which should take up enough time to execute.

React should take advantage of this.

**Comment by 132yse** [comment]
@mindplay-dk Hey, I'm back. I fixed the previous bug. Now the use case is working well.
It confirm that my hypothesis about Message Channel is correct and Fre's schedule thinking is correct now.
How amazing!
demo [here](https://github.com/132yse/fre/tree/master/demo/concurrent), you can run `npm concurrent` to see it.

**Comment by mindplay-dk** [comment]
Congratulations! 🙌

I've been testing it in my sandbox project, and it seems to work nicely! 😀

I wanted to know the difference between message channel, `setTimeout`, `requestIdleCallback` and the `Promise.resolve()` technique in terms of scheduling.

To test this, try the following...

In `scheduler.js`, replace the two `port.postMessage(null)` statements with `planWork()` and add this:

```js
const defer = Promise.prototype.then.bind(Promise.resolve());
//TODO
function planWork() {
  //port.postMessage(null);
  setTimeout(portMessage, 0);
  //requestIdleCallback(portMessage);
  //defer(portMessage);
}
```

Now you can compare each of the 4 techniques by uncommenting each line - here's what I found:

  * `setTimeout(portMessage, 0)` works just as well as the message channel.
  * `requestIdleCallback(portMessage)` also works just the same.
  * `defer(portMessage)` is *really* bad - don't use this!

As far as I can tell, you can safely replace the message channel code with just a plain `setTimeout` or `requestIdleCallback`. (I expect both of these, as well as the message channel dispatch, very likely all execute right after each other in the event loop.)

Even with very high number of components, and higher blocking times per component, I don't see a difference between the message channel technique and `setTimeout` - they both give super smooth results!

Probably the safest bet is `setTimeout`, which is a much older browser feature, so that would be my recommendation.

It'll save a few lines of code as well. 😉

**Comment by 132yse** [comment]
`setTimeOut` can't run executed 0 ms. In fact, it run executed at least 10 ms, which will cause a waste of time. 
`requestIdleCallback `  has less compatible and has low priority, which can be affected by other tasks.
so……Message Channel is the best.

**Pull request merged**